src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParser.cs

```diff
@@ -147,14 +147,38 @@ private TypeNameParser(ReadOnlySpan<char> name, bool throwOnError, TypeNameParse
             {
                 _inputString = capturedBeforeProcessing;
             }
+            else
+            {
+                // Every constructed generic type needs the generic type definition.
+                if (!TryDive(_parseOptions, ref recursiveDepth))
+                {
+                    return null;
+                }
+                // If that generic type is a nested type, this needs to be taken into account. Example:
+                // "Namespace.Declaring+NestedGeneric`1[GenericArg]" requires 5 TypeName instances:
```

Comment:
`Namespace.Declaring+NestedGeneric`1[GenericArg]` should only require 4 TypeName instances.

I think this comment is not correct. Also, the implementation of GetNodeCount seems to have a bug - it is double-counting nested TypeName instances.

Reply:
> `Namespace.Declaring+NestedGeneric[GenericArg]` should only require 4 TypeName instances.

Which node from the list below should not be included in your opinion?

```log
- constructed generic type: Namespace.Declaring+NestedGeneric`2[GenericArg] (+1, handled by ParseNextTypeName)
  - declaring type: Namespace.Declaring (+1, handled by TryGetTypeNameInfo)
  - generic type definition: Namespace.Declaring+NestedGeneric`1 (+1, handled by the TryDive above)
     - declaring type: Namespace.Declaring (+1, handled by the if below)
  - generic arguments:
    - simple: GenericArg (+1, handled by ParseNextTypeName)
```

We define it as the total number of `TypeName` instances required to represent it:

https://github.com/dotnet/runtime/blob/89839049b486bf48239dd6faa722defd3de320cb/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs#L275-L276

And we have 5 instances.

![image](https://github.com/user-attachments/assets/32093c45-d5e4-4acf-aff7-0dfeb1904d21)

The declaring type of the generic type definition and of the constructed generic type is the same object:

![image](https://github.com/user-attachments/assets/4e3103c2-d9a5-40db-929d-244ca672cc95)

Do you suggest that because it's the same object, it should not be included?

> Also, the implementation of GetNodeCount seems to have a bug - it is double-counting nested TypeName instances.

It does that on purpose: to include the nodes from generic type definition.

Reply:
> Do you suggest that because it's the same object, it should not be included?

Yes. If it is the same object, it is the same instance and it should be only counted once. 

We have multiple ways to get to that instance through the object model. I do not think that it is the reason to count it multiple times.

Reply:
> Yes. If it is the same object, it is the same instance and it should be only counted once

You are right. I've changed the implementation and the tests, PTAL.